honeybee/cli/edit.py

```diff
@@ -0,0 +1,292 @@
+        tol = parsed_model.tolerance
+
+        # generate the windows for all walls of rooms
+        for room in model.rooms:
+            for face in room.faces:
+                if isinstance(face.boundary_condition, Outdoors) and \
+                        isinstance(face.type, Wall):
+                    face.apertures_by_ratio(ratio, tol)
+
+        # write the new model out to the file or stdout
+        output_file.write(json.dumps(parsed_model.to_dict()))
+    except Exception as e:
+        _logger.exception('Model windows by ratio failed.\n{}'.format(e))
+        sys.exit(1)
+    else:
+        sys.exit(0)
+
+
+@edit.command('windows-by-ratio-rect')
+@click.argument('model-json', type=click.Path(
+    exists=True, file_okay=True, dir_okay=False, resolve_path=True))
+@click.argument('ratio', type=float)
+@click.argument('aperture-height', type=float)
+@click.argument('sill-height', type=float)
+@click.argument('horizontal-separation', type=float)
+@click.option('--vertical-separation', help='An optional number to create a single '
+              'vertical separation between top and bottom apertures.',
+              type=float, default=0, show_default=True)
+@click.option('--output-file', help='Optional file to output the Model JSON string'
+              ' with windows. By default it will be printed out to stdout',
+              type=click.File('w'), default='-')
+def windows_by_ratio_rect(model_json, ratio, aperture_height, sill_height,
+                          horizontal_separation, vertical_separation, output_file):
+    """Add apertures to all outdoor walls of a model given a ratio.\n
+    Note that this method removes any existing apertures and doors from the Walls.\n
+    Any rectangular portions of walls will have customized rectangular apertures
+    using the various inputs.\n
+    \n
+    Args:
+        model_json: Full path to a Model JSON file.\n
+        ratio: A number between 0 and 1 (but not perfectly equal to 1)
+            for the desired ratio between wall area and face area.
+        aperture_height: A number for the target height of the output apertures.
+            Note that, if the ratio is too large for the height, the ratio will
+            take precedence and the actual aperture_height will be larger
+            than this value.
+        sill_height: A number for the target height above the bottom edge of
+            the rectangle to start the apertures. Note that, if the
+            ratio is too large for the height, the ratio will take precedence
+            and the sill_height will be smaller than this value.
+        horizontal_separation: A number for the target separation between
+            individual aperture centerlines.  If this number is larger than
+            the parent rectangle base, only one aperture will be produced.
+    """
+    try:
+        assert os.path.isfile(model_json), 'No JSON file found at {}.'.format(model_json)
```

Comment:
You don't need this check. `Path(exists=True)` does this check for you.

Reply:
Ah, very good point. I forgot that I have this relic from before I used all that the click decorator offers.  I'll fix it in a few places now.

Reply:
Done!